sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/HadoopFsRelation.scala

```diff
@@ -72,7 +72,8 @@ case class HadoopFsRelation(
     val compressionFactor = sqlContext.conf.fileCompressionFactor
     val defaultSize = (location.sizeInBytes * compressionFactor).toLong
     location match {
-      case cfi: CatalogFileIndex if sparkSession.sessionState.conf.fallBackToHdfsForStatsEnabled =>
+      case cfi: CatalogFileIndex if sparkSession.sessionState.conf.fallBackToHdfsForStatsEnabled
+        && defaultSize == sqlContext.conf.defaultSizeInBytes =>
```

Comment:
nit: 
```
      case cfi: CatalogFileIndex if sparkSession.sessionState.conf.fallBackToHdfsForStatsEnabled &&
        defaultSize == sqlContext.conf.defaultSizeInBytes =>
```

Reply:
I have 2 questions here:
1. Is this [defaultSize](https://github.com/apache/spark/pull/25460/files#diff-fcb68cd3c7630f337ce9a3b479b6d0c4R73) the correct data size?
2. If this [defaultSize](https://github.com/apache/spark/pull/25460/files#diff-fcb68cd3c7630f337ce9a3b479b6d0c4R73) is the correct data size. Could we do some benchmark about `(location.sizeInBytes * compressionFactor).toLong` and `CommandUtils.getSizeInBytesFallBackToHdfs(sparkSession, new Path(cfi.table.location), defaultSize)`?

Reply:
@wangyum 
default size is coming as Long.MAX. If it is correct size it will not fallback to HDFS. Even if it is correct size, fallbackhdfs also give the same result. Also it is not a performance sensitive path I guess. Because, it will come here only when it requires to compute statistics, for eg: during join operation. And if the table already has statistics, flow will not come here.

Reply:
Please see benchmark here: https://github.com/apache/spark/pull/24715#issuecomment-521611355

Reply:
@wangyum (location.sizeInBytes * compressionFactor).toLong is always 8.0EB, even after the PR #24715 
I am not sure I understand your comment.  If the statistics doesn't exists, it has to fallback to HDFS. right? Next time onwards it will read from stats cache.

Number of times falling back to HDFS after this PR and #24715 are also same. right?

Reply:
We can avoid this when constructing this `CatalogFileIndex`:
https://github.com/apache/spark/blob/c30b5297bc607ae33cc2fcf624b127942154e559/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L383-L387

Reply:
btw, not `defaultSize == sqlContext.conf.defaultSizeInBytes` but `location.sizeInBytes == sqlContext.conf.defaultSizeInBytes`?
See: https://github.com/apache/spark/pull/22502#discussion_r313938009

But, does this comparison work well even when `sqlContext.conf.defaultSizeInBytes` changed by users?

Reply:
Ah good point! Basically there is no way to tell the table stats are available or not at this point. `sqlContext.conf.defaultSizeInBytes` is configurable and it's possible that the table stats just equal to `sqlContext.conf.defaultSizeInBytes`.

#24715 seems to be able to fix it.

Reply:
Yea, so how about closing this and moving #24715 for more discussions about solving this case?

Reply:
If the table statistics is available here,
https://github.com/apache/spark/blob/0ea8db9fd3d882140d8fa305dd69fc94db62cf8f/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L387-L390,

Then it should be available here too right?
https://github.com/apache/spark/blob/0ea8db9fd3d882140d8fa305dd69fc94db62cf8f/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L43-L45

So, ideally the flow shouldn't come here if the table statistics already exists. That is why even after #24715, `location.sizeInBytes` is 8.0EB